### PR TITLE
SNOW-1458148: Implement Index.equals

### PR DESCRIPTION
**Pull request by sfc-gh-nkumar**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1458148

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Added distributed support for Index.equals method.

### CODE REVIEW

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 1]
```diff
@@ -15589,7 +15589,52 @@ def pct_change(
...
+        Args:
+            other: Snowflake query compiler to compare against.
+
+        Returns:
+            True if self index is equal to other index, False otherwise.
+        """
+        if self._modin_frame.num_index_columns != other._modin_frame.num_index_columns:
+            return False
+
+        # Join on row position columsn to compare order of data.
```
Comment (sfc-gh-vbudati) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:15608]:
```suggestion
        # Join on row position columns to compare order of data.
```
nit

**Review by sfc-gh-vbudati — APPROVED** [review]
Looks great, thanks Naresh!

File: CHANGELOG.md [hunk 2]
```diff
@@ -77,6 +77,7 @@
 - Added support for `DataFrame.corr`.
 - Added support for `DataFrame.equals` and `Series.equals`.
 - Added support for `DataFrame.reindex` and `Series.reindex`.
+- Added support for `Index.equals`.
```
Comment (sfc-gh-helmeleegy) [inline CHANGELOG.md:80]: I think this needs to be moved to the new release above.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 3]
```diff
@@ -15589,7 +15589,54 @@ def pct_change(
                 ).frame
             )
 
-    def equals(self, other: "SnowflakeQueryCompiler") -> "SnowflakeQueryCompiler":
+    def index_equals(self, other: "SnowflakeQueryCompiler") -> bool:
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:15592]: Aren't we keeping docstrings under the `docstrings` folder for Index methods? cc: @sfc-gh-vbudati

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 4]
```diff
@@ -15589,7 +15589,54 @@ def pct_change(
...
+        * The order of the elements inside the Index object.
+
+        Args:
+            other: Snowflake query compiler to compare against.
+
+        Returns:
+            True if self index is equal to other index, False otherwise.
+        """
+        if self._modin_frame.num_index_columns != other._modin_frame.num_index_columns:
+            return False
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:15606]: It seems we need a test covering this case when the number of index columns is not matching.

Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:15606]: Ah, this needs to go away. I later learned that currently we only support lazy index for single index.

File: CHANGELOG.md (hunk 2 above)
Comment (sfc-gh-nkumar) [reply CHANGELOG.md:80]: good catch! moved.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 3 above)
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:15592]: No, that change has not been made yet @sfc-gh-helmeleegy. All index docstrings are currently in the same file as the code.

Comment (sfc-gh-helmeleegy) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:15592]: I see. Thanks.

**Review by sfc-gh-helmeleegy — APPROVED** [review]
LGTM, thanks Naresh!

**Review by sfc-gh-evandenberg — APPROVED** [review]
LGTM

**Pull request merged**